### PR TITLE
Enable option use /etc/sysconfig/rsyslog

### PR DESCRIPTION
**Pull request by portante**
Add an option to use `/etc/sysconfig/rsyslog` in the systemd service file.
This lets Fedora drop the patch they are carrying, and is probably
useful for other platforms to consider.

### CODE REVIEW

**Comment by Whissi** [comment]
I am against this PR: */etc/sysconfig* is a distribution specific directory. Do we want to add something similar for Debian? ArchLinux?

Well, we could make this configurable so package maintainers could set the path for their systems.

But isn't a config file in systemd world a bad thing? I am not a systemd expert but in Gentoo we warn about that:

> It should be noted that Environment= and EnvironmentFile= keys were not designed to make unit files configurable but to allow exporting environment variables needed for programs. Therefore, all variables declared there are exported into the environment of spawned program. This may result in exposing configuration data to third parties. 

Instead I was told that one good thing in systemd is the possibility to overwrite units. See https://lwn.net/Articles/584175/ and https://wiki.archlinux.org/index.php/Systemd#Editing_provided_units.  It would surprise me if this concept should not be taken over by RH-based distributions.

**Comment by portante** [comment]
Yes, it is a distribution specific directory.

We could remove `rsyslog.service.in` from the source tree and move it the https://github.com/rsyslog/rsyslog-pkg-rhel-centos repo?  What do folks think?

That said, this patch does not change the existing behavior, and the use of the `EnvironmentFile=` line is only there if explicitly requested.

But the source tree carries a number of distribution specific behaviors, see `configure.ac`,  so do you see a precedent for not carrying this kind of change?

Regarding the proper us of the `systemd` `Environment` and `EnvironmentFile` keys in `.service` files: Fedora is shipping the use of `/etc/sysconfig/rsyslog` to affect options for rsyslog today.  So it would seem that this is a behavior that needs to be taken up with Fedora, CentOS, and RHEL for how they use options like that.  It is fairly pervasive, though, as there are many other packages that do the same thing.  If we want to engage here in being part of the solution were rsyslog reads environment variables to affect its operation (like it already does for debugging, which I have used many times), I'm all for it.  But I suspect that is low on the totem pole for work that folks need.

**Comment by rgerhards** [comment]
@mbiebl It would be great if you could cast your ballot.

**Comment by mbiebl** [comment]
I'm against it as well. /etc/sysconfig is a fedora specific concept and we want to get rid of /etc/sysconfig (redhat) and /etc/default (debian) as much as possible.
systemd provides enough methods to override options. I don't see why we need an external file here via EnvironmentFile=

@portante Which options exactly do you need to pass to rsyslogd which can't be configured via rsyslog.conf?

File: rsyslog.service.in [hunk 1]
```diff
@@ -6,7 +6,9 @@ Documentation=http://www.rsyslog.com/doc/
 
 [Service]
 Type=notify
-ExecStart=@sbindir@/rsyslogd -n
+@syslogd_options_env_file@
+ExecStart=@sbindir@/rsyslogd -n@syslogd_options@
+UMask=0066
```
Comment (mbiebl) [inline rsyslog.service.in:11]: You slipped an unrelated change into this pull request. The UMask change I mean.

**Comment by rgerhards** [comment]
What would be the best way to handle this? Would it make sense to move this to the packaging projects and drop systemd config files? When we introduced them, systemd was new. Nowadays, I guess each distro has its own unit files and I guess there is little need for this inside rsyslog. But I really don't know...

**Comment by mbiebl** [comment]
no, please don't drop the systemd unit files. They work fine as is and we use them unmodified in Debian.

**Comment by Whissi** [comment]
Same for Gentoo, please don't drop. Keep shipping a good default unit file.

**Comment by rgerhards** [comment]
ok, than I think it is up to you maintainers to reach some agreement. @theinric @darix want to step in as well?

**Comment by mbiebl** [comment]
If Fedora desperately needs the /etc/sysconfig functionality, then they should patch that in.

Upstream (systemd) is pretty clear that EnvironmentFile= should be avoided as much as possible.

If there is a real need for passing command line arguments to the rsyslogd daemon, what you can do today is override the ExecStart line via a drop-in or by copying the rsyslog.service file to /etc/systemd/system.

systemctl edit (--full) rsyslog.service makes that quite convenient.

Overriding the full ExecStart= line does indeed have some downsides, so maybe as a middle ground, we could have

```
ExecStart=/usr/sbin/rsyslogd -n $OPTS
```

Then the admin can very easily pass additional options via systemctl edit rsyslog.service and adding
```
[Unit]
Environement=OPTS=--my-args
```

Fedora then would need to do a one-time migration of the existing /etc/sysconf/rsyslog file and then drop support for it.

The big advantage of overrides, that tools like systemd-delta will should you those modifications directly.

**Comment by portante** [comment]
@mbiebl, how is this change making any impact on the other distributions? You have to specifically enable a switch to get the behavior, unless something is broken with the patch, right?  Or are you talking about the UMask additional line?

And we carry all kinds of distribution specific behaviors in the base here.

If folks want to maintain it as it was, I'll drop this patch.  No sweat.

The UMask change was to sync what was in Fedora so that rsyslog does not write group and other readable/writable files.  I thought that would be desirable for other distributions, but apparently not.  Sorry for creating any problems here.

**Comment by mbiebl** [comment]
@portante that's the wrong question. The question is, whether we should pull in distro specifics if we can avoid it because we have better options.
See @Whissi  remark earlier. Should we start add EnvironmentFile=/etc/default/rsyslog, /etc/conf.d/rsyslog and whatever other distros might use out there?

systemd brings some long needed unification for the Linux base system. We shouldn't work against that

**Comment by mbiebl** [comment]
and to answer your question specifially: It does indeed have impact on other distros, since now the configuration of the daemon becomes distro specific again, for no good reason.

**Comment by portante** [comment]
@mbiebl, so right now Fedora has one build of the systemd configuration file, RHEL keeps its own, based off of Fedora, and CentOS follows RHEL.  So those are three different distributions that use the same configuration.  However, Adiscon also puts out its own set of Fedora and EPEL packages (for use on CentOS and RHEL) which are different and don't carry the same behaviors as rsyslog shipped by Fedora and company.  So should we copy this Fedora/RHEL/CentOS behavior to yet a fourth location, or  should we keep it in one place so that we reduce the behavioral differences?

My vote is for an upstream package to help the various downstream distros "get it right" rather than be dogmatic and rigid in its approach.

It seems we are all for getting Fedora/RHEL/CentOS to change how they are using systemd for rsyslog, but given how long that takes, and **how these distro differences in behavior for rsyslog has caused problems for users**, it also seems worth while to address it here to help rsyslog itself.  And we should probably keep in mind that the Fedora supported versions 'age' out quicker than what rsyslog-pkg-rhel-centos is maintaining, as it builds for older distribution versions than what RHEL and CentOS support.  So by unifying it we simplify the support long term as well.

It was hoped that this change struck a balance between impact to other distributions and unifying existing behaviors:
 * It does not affect other distros, evidenced by the fact that while the `rsyslog.service.in` file is changed, the resulting `rsyslog.service` file generated by the build process is identical as before for those distributions, nothing should have to change there (we assert; ignore the `UMask` change for now)
 * The logic is kept in one place to help bridge the behavioral differences across the different distributions

I am not seeing the objective way this change affects other distros, and I'd like to learn what I am missing, as perhaps those distros do their own processing of `rsyslog.service.in` instead of using the build process?

Having said all this, if that is not sufficient, let's drop this change, and see if the existing patch to unify rsyslog-pkg-rhel-centos with Fedora is worth to do (see https://github.com/rsyslog/rsyslog-pkg-rhel-centos/issues/19).

**Comment by mbiebl** [comment]
Doing it right means, we don't add distro specific, legacy stuff for dubious reasons.

Rainer wanted my opinion, my opinion is that adding such distro hacks is bad and becomes unmaintaintable over time. The best practice is to not use /etc/sysconfig files under systemd but use the systemd provided mechanisms
It seems to me that you are trying desperately to translate the old sysv init script 1to1 to systemd.
You should use this opportunity to get rid of /etc/sysconfig/rsyslog, not prolong its existence.

**Comment by darix** [comment]
1. environment file should not be used
2. you dont have to edit systemd unit files.

see: https://discourse.nordisch.org/t/per-service-ulimits/374 and https://discourse.nordisch.org/t/apparmor-and-multiple-instances-of-the-same-service/373
